crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,648 @@
+        }
+        "bool" => to_json::<bool>(data),
+        "Argument" => to_json::<Argument>(data),
+        "CallArg" => to_json::<CallArg>(data),
+        "ObjectDigest" => to_json::<ObjectDigest>(data),
+        "ObjectID" => to_json::<ObjectID>(data),
+        "ProgrammableTransaction" => to_json::<ProgrammableTransaction>(data),
+        "SequenceNumber" => to_json::<SequenceNumber>(data),
+        "SuiAddress" => to_json::<SuiAddress>(data),
+        "String" => to_json::<String>(data),
+        "TransactionData" => to_json::<TransactionData>(data),
+        "TransactionKind" => to_json::<TransactionKind>(data),
+        unsupported => bail!("Requested type {} is currently unsupported.", unsupported),
+    };
+    match json_obj {
+        Ok(json) if pretty => Ok(serde_json::to_string_pretty(&json)?),
+        Ok(json) => Ok(serde_json::to_string(&json)?),
+        Err(e) => bail!("Cannot convert bcs to json, {e}"),
+    }
+}
+
+fn internal_bcs_from_json(typename: &str, data: &str) -> Result<Vec<u8>, Error> {
+    let type_ = if typename.starts_with("vector") {
+        "vector"
+    } else {
+        typename
+    };
+
+    let bcs_bytes = match type_ {
+        "u8" | "u16" | "u32" | "u64" | "u128" | "u256" => string_integer_to_bcs(typename, data),
+        "bool" => to_bcs::<bool>(data),
+        "vector" => to_bcs_from_vector(typename, data),
+        "Argument" => to_bcs::<Argument>(data),
+        "CallArg" => to_bcs::<CallArg>(data),
+        "ObjectDigest" => to_bcs::<ObjectDigest>(data),
+        "ObjectID" => to_bcs::<ObjectID>(data),
+        "ProgrammableTransaction" => to_bcs::<ProgrammableTransaction>(data),
+        "SequenceNumber" => to_bcs::<SequenceNumber>(data),
+        "SuiAddress" => to_bcs::<SuiAddress>(data),
+        "String" => to_bcs::<String>(data),
+        "TransactionData" => to_bcs::<TransactionData>(data),
+        "TransactionKind" => to_bcs::<TransactionKind>(data),
+        unsupported => bail!("Requested type {} is currently unsupported.", unsupported),
+    };
+
+    match bcs_bytes {
+        Ok(bcs) => Ok(bcs),
+        Err(e) => bail!("Cannot convert json to bcs, {}", e),
+    }
+}
+
+/// Return 0 for success and 1 for failure
+///
+/// Converts the JSON data into a BCS array.
+/// The result points to the address where the new BCS
+/// array is stored. Don't forget to deallocate the memory
+/// by calling the sui_bcs_json_free_array function.
+#[ffi_export]
+fn sui_bcs_from_json(
+    type_name: char_p::Ref<'_>,
+    json_data: char_p::Ref<'_>,
+    result: *mut *const u8,
+    length: *mut *mut size_t,
+) -> usize {
+    match internal_bcs_from_json(type_name.to_str(), json_data.to_str()) {
+        Ok(output) => {
+            let mut vec = output;
+            vec.shrink_to_fit();
+            assert!(vec.len() == vec.capacity());
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            std::mem::forget(vec); // prevent deallocation in Rust
```

Comment:
You could also follow this [example](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#examples-3) and use ManuallyDrop.